### ISSUE TITLE
Update points array inplace does not trigger ModifiedEvent

### ISSUE THREAD

**Issue by banesullivan** [pyvista/pyvista#1262]
The following would have you believe that updating an array associated with a mesh inplace would trigger a modified event but it does not:

https://github.com/pyvista/pyvista/blob/6895b14c8373cc7566af70e5994375475d1d749f/pyvista/core/pyvista_ndarray.py#L46-L55

Here is an example:

```py
from pyvista import examples

mesh = examples.load_hexbeam()

import vtk as _vtk
mesh.AddObserver(_vtk.vtkCommand.ModifiedEvent, lambda *args: print('fooooo'))
```

Call modified and see the callback
```
>>> mesh.Modified()
fooooo
```

Update the points inplace which should call modified, but we do not see the callback output 😞 
```py
>>> mesh.points[:,-1] *=0.5
```

**Comment by banesullivan** [comment]
@pyvista/developers, can we get this resolved before #1257 ?

**Comment by akaszynski** [comment]
> @pyvista/developers, can we get this resolved before #1257 ?

Yes.  Adding to milestone.

**Comment by akaszynski** [comment]
I've become quite perplexed, because the VTK object does indeed get modified.  Placing logging into `__setitem__` shows that the original VTK array is modified.  Additionally, upstream objects are modified (as demonstrated by the `BackgroundPlotter`).  Update occurs instantly, even when setting `auto_update=0.0`:
![tmp](https://user-images.githubusercontent.com/11981631/114627029-cd07a580-9c71-11eb-8950-a5349aa58fe9.png)

Is there a specific bug you're looking to solve with this?  While I agree that it's strange that `mesh.points[:] *= 0.5` isn't triggering an update, it seems that it's not necessary (at least for the plotter).

**Comment by banesullivan** [comment]
I originally discovered this issue while looking into https://github.com/Kitware/ipyvtk-simple/issues/29

To resolve it, I added an observer on the renderer to look for modified events. This approach worked for that particular issue where the camera was modified, but I want to also be able to adjust the mesh and/or its point/cell data and have that immediately update in the ipyvtk canvas. That's when I noticed a modified event wasn't being triggered.

**Comment by banesullivan** [comment]
Same effect with array assignment: `mesh['sample_point_scalars'] = np.zeros(mesh.n_points)` (no modified event)

**Comment by banesullivan** [comment]
Issue to resolve here:

```py
import pyvista as pv
from pyvista import examples

pv.set_jupyter_backend("ipyvtklink") # or ipyvtk_simple

mesh = examples.load_hexbeam()
p = pv.Plotter()
p.add_mesh(mesh, show_edges=True)
p.show()
```

Then run

```py
mesh['sample_point_scalars'] = np.zeros(mesh.n_points)
```

And notice that the canvas is not updated until interacting with it (or after the render timer)

**Comment by akaszynski** [comment]
```
import numpy as np
import pyvista as pv
from pyvista import examples

# pv.set_jupyter_backend("ipyvtklink") # or ipyvtk_simple

mesh = examples.load_hexbeam()

from pyvistaqt import BackgroundPlotter
p = BackgroundPlotter(auto_update=0.0)
p.add_mesh(mesh, show_edges=True)

# wait and then modify

mesh['sample_point_scalars'] = np.zeros(mesh.n_points)
mesh.Modified()
mesh.GetPoints().Modified()
mesh.GetPointData().Modified()
mesh.GetPointData().GetScalars().Modified()
```

This may be a VTK issue, because I'm not able to get to update it regardless of how I try to trigger the pipeline unless I force it with `p.render()`

**Comment by whophil** [comment]
In the OP , the `ModifiedEvent` observer was placed on the dataset, but the array itself was directly modified. I don't know if VTK is supposed to do this internally, but we can also explicitly call `Modified()`on that dataset when the VTKArray is updated, as in:

```python
    def __setitem__(self, key: int, value):
        """Implement [] set operator.
        When the array is changed it triggers "Modified()" which updates
        all upstream objects, including any render windows holding the
        object.
        """
        super().__setitem__(key, value)
        if self.VTKObject is not None:
            self.VTKObject.Modified()

        # <-- NEW STUFF
        # the associated dataset should also be marked as modified
        dataset = self.dataset.Get()
        if dataset is not None:
            dataset.Modified()
        # --> 
```
You can see the change here: https://github.com/whophil/pyvista/commit/077f9852cb40302942843105993aa0207ccdc313

Not sure if this will fix rendering issues.

**Comment by whophil** [comment]
Relevant discussion here: https://vtk.org/pipermail/vtkusers/2016-March/094601.html

> The polydata->Modified() call will tell the mapper that the content of the poly data has changed. Direct changes to the point and cell data structures do not trigger the Modified() call on the poly data automatically. vtkPolyData could be made to do that by observing its point and cell data objects, but it does not, so you must do it manually.
>
> I hope that clears things up for you,
>Cory

**Comment by akaszynski** [comment]
Nice work @whophil, this was really annoying me.

**Comment by akaszynski** [comment]
Fixed in #1315
